docs/IPLD-ProllyTree-Spec.md

```diff
@@ -0,0 +1,271 @@
+
+![img7](./images/spec/img_7.png)
+
+Framework is used for tree building. We input sorted data into the framework then call the BuildTree to get a new ProllyTree from the data.The tree is built level by level(bottom) by LevelBuilders. Leaf nodes are linked by parent nodes by the cid link(IPLD node). All nodes are saved in NodeStore referenced by cid.
+
+![img8](./images/spec/img_8.png)
+
+IPLD ProllyTree implemented the IPLD ADL(Advanced Layout) interface based on above structures.It includes ProllyTreeADLPrototype, Builder, TreeAssembler(MapAssembler) and ADL Node. It helps people who want to build large map(dict) by ProllyTree with ADL in IPLD world.
+
+## Modules
+### Storge(NodeStore)
+
+```golang
+type NodeStore interface {
+	WriteNode(ctx context.Context, nd *ProllyNode, prefix *cid.Prefix) (cid.Cid, error)
+	ReadNode(ctx context.Context, c cid.Cid) (*ProllyNode, error)
+
+	WriteTree(ctx context.Context, tree *ProllyTree, prefix *cid.Prefix) (cid.Cid, error)
+	ReadTree(ctx context.Context, c cid.Cid) (*ProllyTree, error)
+
+	WriteTreeConfig(ctx context.Context, cfg *TreeConfig, prefix *cid.Prefix) (cid.Cid, error)
+	ReadTreeConfig(ctx context.Context, c cid.Cid) (*TreeConfig, error)
+
+	WriteProof(ctx context.Context, prf Proof, prefix *cid.Prefix) (cid.Cid, error)
+	ReadProof(ctx context.Context, c cid.Cid) (Proof, error)
+
+	LinkSystem() *ipld.LinkSystem
+
+	Close()
+}
+```
+
+NodeStore is an import module that saves almost all information and data of ProllyTree. It is used anywhere in the project such as tree building, updating, getting…… We can also read the saved tree before. The nodestore we use in the project is based on the blockstore(ipfs)
+### ProllyTree
+```golang
+type ProllyTree struct {
+ProllyRoot
+root       ProllyNode
+ns         NodeStore
+treeConfig TreeConfig
+treeCid    *cid.Cid
+
+mutating  bool
+mutations *Mutations
+}
+```
+ProllyTree is the core part in the project, it provides api for people to use the tree, it can also play the role as a package like in ADL, IPLD-Prolly-Indexer. The root node(ProllyNode) is the entry of the tree. When we need traverse the tree or some branch path in tree, we read the cidlink in node value and get relative node from NodeStore by cid.TreeConfig contains important information about the prollytree such as node size limit, hash function, cid version, codec and so on.It is set while building from framework.Mutations is used for tree mutating(updating), we won’t update for every mutating action such as putting(includes inserting and modifying), deleting.Mutations caches these actions then apply them while Rebuilding is called.
```

Comment:
```suggestion
ProllyTree is the core part in the project, it provides api for people to use the tree, it can also play the role as a package like in ADL, IPLD-Prolly-Indexer. The root node(ProllyNode) is the entry of the tree. When we need traverse the tree or some branch path in tree, we read the cidlink in node value and get relative node from NodeStore by cid.TreeConfig contains important information about the prollytree such as node size limit, hash function, cid version, codec and so on.It is set while building from framework.Mutations is used for tree mutating(updating), we won’t update for every mutating action such as putting(includes inserting and modifying), deleting. Mutations caches these actions then apply them when calling `Rebuild()`.
```